File: week-5/lecture/4-merge-sort/merge-sort.cpp
/**
 * @file merge-sort.cpp
 * @brief C++ version of the week 5 merge-sort lecture example.
 */

#include <iostream>
#include <vector>

using namespace std;

/**
 * @brief Merge two sorted lists into one sorted list.
 * @param left_half const vector<int>& Sorted left half.
 * @param right_half const vector<int>& Sorted right half.
 * @return vector<int> Merged list in ascending order.
 */
vector<int> merge(const vector<int>& left_half, const vector<int>& right_half) {
    vector<int> merged;
    size_t left_index = 0;
    size_t right_index = 0;

    while (left_index < left_half.size() && right_index < right_half.size()) {
        if (left_half[left_index] <= right_half[right_index]) {
            merged.push_back(left_half[left_index++]);
        } else {
            merged.push_back(right_half[right_index++]);
        }
    }

    merged.insert(merged.end(), left_half.begin() + static_cast<long>(left_index), left_half.end());
    merged.insert(merged.end(), right_half.begin() + static_cast<long>(right_index), right_half.end());
    return merged;
}

/**
 * @brief Sort the list in ascending order using merge sort.
 * @param array const vector<int>& Input list of integers.
 * @return vector<int> Sorted copy of array.
 */
vector<int> mergeSort(const vector<int>& array) {
    if (array.size() <= 1) {
        return array;
    }

    size_t middle = array.size() / 2;
    vector<int> left_half(array.begin(), array.begin() + static_cast<long>(middle));
    vector<int> right_half(array.begin() + static_cast<long>(middle), array.end());
    vector<int> merged = merge(mergeSort(left_half), mergeSort(right_half));

    for (int value : merged) {
        cout << value << ' ';
    }
    cout << '\n';
    return merged;
}

/**
 * @brief Run the merge-sort lecture example.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> data = {6, 5, 3, 1, 8, 7, 2, 4};
    vector<int> sorted = mergeSort(data);
    cout << "Sorted array: ";
    for (int value : sorted) {
        cout << value << ' ';
    }
    cout << '\n';
    return 0;
}

File: week-5/lecture/4-merge-sort/merge-sort.py
"""Merge Sort example.

This file belongs to the week 5 merge sort lecture example. It recursively
splits the list and then merges the sorted halves.
"""


def merge(left_half, right_half):
    """Merge two sorted lists into one sorted list.

    Args:
        left_half (object): The left_half input used by this function.
        right_half (object): The right_half input used by this function.

    Returns:
        list: The list produced by the algorithm.
    """
    merged = []
    left_index = 0
    right_index = 0

    while left_index < len(left_half) and right_index < len(right_half):
        if left_half[left_index] <= right_half[right_index]:
            merged.append(left_half[left_index])
            left_index += 1
        else:
            merged.append(right_half[right_index])
            right_index += 1

    merged.extend(left_half[left_index:])
    merged.extend(right_half[right_index:])
    return merged


def mergeSort(array):
    """Sort the list in ascending order using merge sort.

    Args:
        array (list): The list being processed by the algorithm.

    Returns:
        list: The list produced by the algorithm.
    """
    if len(array) <= 1:
        return array

    middle = len(array) // 2
    left_half = mergeSort(array[:middle])
    right_half = mergeSort(array[middle:])
    merged = merge(left_half, right_half)
    print(merged)
    return merged


def main():
    """Run the small demonstration for this teaching example.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    data = [6, 5, 3, 1, 8, 7, 2, 4]
    print("Sorted array:", mergeSort(data))


if __name__ == "__main__":
    main()
